### PR TITLE
Implementing state manager

### PR DESCRIPTION
**Pull request by iambibhas**

### CODE REVIEW

File: hasjob/models/jobpost.py [hunk 1]
```diff
@@ -228,7 +228,8 @@ def expiry_date(self):
     def after_expiry_date(self):
         return self.expiry_date + timedelta(days=1)
 
-    state.add_conditional_state('OLD', state.LISTED, lambda proposal: proposal.datetime < datetime.utcnow() - agelimit, label=('old', __("Old")))
```
Comment (jace) [inline hasjob/models/jobpost.py]: This is correct, but it's a jobpost, not a proposal. OLD is a conditional state for LISTED.

Comment (iambibhas) [reply hasjob/models/jobpost.py]: @jace right. typo. I removed it thinking OLD should not be a state. but yeah, it is only a conditional state for LISTED. so it makes sense. I'll revert it back.

File: hasjob/views/listing.py [hunk 2]
```diff
@@ -721,31 +715,33 @@ def confirm_email(domain, hashid, key):
...
+                    ).filter(JobPost.state.POSTPENDING).filter(
+                        JobPost.datetime > datetime.utcnow() - timedelta(days=1)
+                    ).count()
                 if post_count > app.config['THROTTLE_LIMIT']:
                     flash(u"We have received too many posts with %s addresses in the last 24 hours. "
                         u"Posts are rate-limited per domain, so yours was not confirmed for now. "
                         u"Please try confirming again in a few hours."
                         % post.email_domain, category='info')
                     return redirect(url_for('index'))
             post.email_verified = True
```
Comment (jace) [inline hasjob/views/listing.py]: Move this line into the `post.confirm` transition method.

Comment (jace) [reply hasjob/views/listing.py]: In fact, all of this logic should move into the `post.confirm` method, but we can do that in the next PR where we remove the logic from views and move it into transitions.

File: runserver.py [hunk 3]
```diff
@@ -8,4 +8,4 @@
     port = int(sys.argv[1])
 except (IndexError, ValueError):
     port = 5000
-app.run('0.0.0.0', port=port)
+app.run('0.0.0.0', port=port, debug=True)
```
Comment (jace) [inline runserver.py]: Remove this. `DEBUG = True` should be specified in `instance/development.py`. We need to disable debug mode sometimes to test some functionality.

File: hasjob/views/listing.py [hunk 4]
```diff
@@ -642,18 +639,15 @@ def reject_post(post, reason, spam=False):
...
     moderateform = forms.ModerateForm()
     if moderateform.validate_on_submit():
-        post.closed_datetime = datetime.utcnow()
         post.review_comments = moderateform.reason.data
-        post.review_datetime = datetime.utcnow()
-        post.reviewer = g.user
-        flashmsg = "This job post has been moderated."
-        post.status = POSTSTATUS.MODERATED
+        post.moderate()
+        flashmsg = post.moderate.data['message']
```
[the remark this answers is not in the document]
Comment (iambibhas) [reply hasjob/views/listing.py]: What if a transition function has multiple transitions? How does that work?

File: hasjob/models/jobpost.py [hunk 5]
```diff
@@ -225,66 +228,47 @@ def expiry_date(self):
...
-    def is_pending(self):
-        return self.status == POSTSTATUS.PENDING
-
-    def is_unpublished(self):
-        return self.status in POSTSTATUS.UNPUBLISHED
-
-    def is_listed(self):
-        now = datetime.utcnow()
-        return (self.status in POSTSTATUS.LISTED) and (
-            self.datetime > now - agelimit)
```
[the remark this answers is not in the document]
Comment (iambibhas) [reply hasjob/models/jobpost.py]: Isn't `NEW` state the same thing? it's a conditional state of PUBLIC and checks for datetime as well.

[the remark this answers is not in the document]
Comment (iambibhas) [reply hasjob/models/jobpost.py]: I'm removing `NEW` and `OLD` and keeping `LISTED` as conditional state. It's either listed or not. no new or old.

File: hasjob/templates/confirm.html.jinja2 [hunk 6]
```diff
@@ -2,7 +2,7 @@
 {% from "baseframe/forms.html.jinja2" import renderfield %}
 {% block title %}{{ post.headline|e }}{% endblock %}
 {% block sheet %}
-  {% if post.is_draft() %}
+  {% if post.state.DRAFT %}
```
[the remark this answers is not in the document]
Comment (iambibhas) [reply hasjob/templates/confirm.html.jinja2]: I've been using `state.DRAFT` everywhere. But yeah. we should decide one. I mixed them up at one point and it wasn't pretty. If we decide on `state.DRAFT`, let's remove the support for `state.is_draft`. Having both available will always be confusing and someone will end up mixing them up.

File: hasjob/views/helper.py [hunk 7]
```diff
@@ -398,12 +398,12 @@ def getposts(basequery=None, pinned=False, showall=False, statuses=None, ageless
...
         pinned = False  # No pinning when browsing archives
     if not statuses:
-        statuses = POSTSTATUS.LISTED
+        statuses = POST_STATE.LISTED
 
     if basequery is None:
         basequery = JobPost.query
 
-    query = basequery.filter(JobPost.status.in_(statuses)).options(*JobPost._defercols).options(db.joinedload('domain'))
+    query = basequery.filter(JobPost._state.in_(statuses)).options(*JobPost._defercols).options(db.joinedload('domain'))
```
[the remark this answers is not in the document]
Comment (iambibhas) [reply hasjob/views/helper.py]: `statuses` here is a list of status IDs I think. Let me recheck.

File: hasjob/views/listing.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jace) [reply hasjob/views/listing.py]: All transitions happen.

[the remark this answers is not in the document]
Comment (jace) [reply hasjob/views/listing.py]: In the case of AbortTransition, all of them should abort.

File: hasjob/models/jobpost.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jace) [reply hasjob/models/jobpost.py]: Fair enough. `LISTED` should suffice.

File: hasjob/templates/confirm.html.jinja2 (hunk 6 above)
[the remark this answers is not in the document]
Comment (jace) [reply hasjob/templates/confirm.html.jinja2]: I’d say these should be the project’s conventions and not enforced by StateManager. In this case, the campaign code needs to change as well.

File: hasjob/views/listing.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (iambibhas) [reply hasjob/views/listing.py]: That I got, I'm asking which transitions message should the transition return?

File: hasjob/templates/confirm.html.jinja2 (hunk 6 above)
[the remark this answers is not in the document]
Comment (iambibhas) [reply hasjob/templates/confirm.html.jinja2]: So we need to choose one of these for Hasjob.

File: hasjob/models/jobpost.py [hunk 8]
```diff
@@ -225,66 +229,49 @@ def expiry_date(self):
...
+        self.review_datetime = db.func.utcnow()
+        self.review_comments = reason
+        self.reviewer = current_auth.user
+
+    @state.transition(state.PUBLIC, state.MODERATED, title=__("Moderate"), message=__("This job post has been moderated"), type='primary')
+    def moderate(self, reason):
+        self.closed_datetime = db.func.utcnow()
+        self.review_datetime = db.func.utcnow()
+        self.review_comments = reason
+        self.reviewer = current_auth.user
```
Comment (jace) [inline hasjob/models/jobpost.py]: Both these should accept `user` as a parameter. Don't assume `current_auth.user`.

File: hasjob/models/jobpost.py [hunk 9]
```diff
@@ -3,18 +3,19 @@
 from collections import defaultdict
 from datetime import datetime, timedelta
 from werkzeug import cached_property
-from flask import url_for, escape, Markup
+from flask import g, url_for, escape, Markup
```
Comment (jace) [inline hasjob/models/jobpost.py]: What is `g` being used for? Importing it in a model is an indication that something is wrong.

File: hasjob/models/jobpost.py [hunk 10]
```diff
@@ -225,66 +228,47 @@ def expiry_date(self):
     def after_expiry_date(self):
         return self.expiry_date + timedelta(days=1)
 
-    def status_label(self):
-        if self.status == POSTSTATUS.DRAFT:
-            return _("Draft")
-        elif self.status == POSTSTATUS.PENDING:
-            return _("Pending")
-        elif self.is_new():
-            return _("New!")
```
[the remark this answers is not in the document]
Comment (jace) [reply hasjob/models/jobpost.py]: We're still missing a conditional state for `NEW` (<24 hours), which is not the same as `OLD` (> 30 days) or `LISTED` (< 30 days).

File: hasjob/models/jobpost.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jace) [reply hasjob/models/jobpost.py]: `NEW` is still required for the 24 hours badge.

File: hasjob/models/jobpost.py [hunk 11]
```diff
@@ -225,66 +229,49 @@ def expiry_date(self):
...
-    def is_unacceptable(self):
-        return self.status in POSTSTATUS.UNACCEPTABLE
-
-    def is_old(self):
-        return self.datetime <= datetime.utcnow() - agelimit
-
-    def pay_type_label(self):
-        return PAY_TYPE.get(self.pay_type)
+    state.add_conditional_state('LISTED', state.PUBLIC, lambda jobpost: jobpost.datetime >= datetime.utcnow() - agelimit, label=('listed', __("Listed")))
+    state.add_conditional_state('CONFIRMABLE', state.UNPUBLISHED, lambda jobpost: jobpost.admin_is(current_auth.user), label=('confirmable', __("Confirmable")))
```
Comment (jace) [inline hasjob/models/jobpost.py]: This one is making me uncomfortable for the `current_auth` import. Ideally, the `admin_is` method should be replaced with an `edit` permission granted in `self.permissions`. This lambda can then become `lambda jobpost: jobpost.current_permissions.edit`

File: hasjob/views/helper.py [hunk 12]
```diff
@@ -397,13 +397,14 @@ def load_viewcounts(posts):
 def getposts(basequery=None, pinned=False, showall=False, statuses=None, ageless=False, limit=2000, order=True):
     if ageless:
         pinned = False  # No pinning when browsing archives
-    if not statuses:
-        statuses = POSTSTATUS.LISTED
+
+    if statuses is None:
+        statuses = JobPost.state.PUBLIC
```
Comment (jace) [inline hasjob/views/helper.py]: We should rename `statuses` to `statusfilter` as that is a better name.

File: hasjob/views/index.py [hunk 13]
```diff
@@ -427,31 +427,31 @@ def index(basequery=None, md5sum=None, tag=None, domain=None, location=None, tit
...
+    return index(basequery=basequery, ageless=True, statuses=[POST_STATE.DRAFT, POST_STATE.PENDING], cached=False)
 
 
 @app.route('/my', methods=['GET', 'POST'], subdomain='<subdomain>')
 @app.route('/my', methods=['GET', 'POST'])
 @lastuser.requires_login
 def my_posts():
     basequery = JobPost.query.filter_by(user=g.user)
-    return index(basequery=basequery, ageless=True, statuses=POSTSTATUS.MY, cached=False)
+    return index(basequery=basequery, ageless=True, statuses=POST_STATE.MY, cached=False)
```
[the remark this answers is not in the document]
Comment (jace) [reply hasjob/views/index.py]: All of these references need to change to filters instead of values.

**Pull request merged**